contracts/protocol/integration/lib/AaveV3.sol

```diff
@@ -0,0 +1,478 @@
+     * @param _asset            The address of the underlying asset borrowed
+     * @param _rateMode         The rate mode that the user wants to swap to
+     *
+     * @return address          Target contract address
+     * @return uint256          Call value
+     * @return bytes            SwapBorrowRate calldata
+     */
+    function getSwapBorrowRateModeCalldata(
+        IPool _lendingPool,
+        address _asset,
+        uint256 _rateMode
+    )
+        public
+        pure
+        returns (address, uint256, bytes memory)
+    {
+        bytes memory callData = abi.encodeWithSignature(
+            "swapBorrowRateMode(address,uint256)", 
+            _asset,
+            _rateMode
+        );
+        
+        return (address(_lendingPool), 0, callData);
+    }
+
+    /**
+     * Invoke to swap borrow rate of SetToken
+     * 
+     * Allows SetToken to toggle it's debt between stable and variable mode
+     * @param _setToken         Address of the SetToken
+     * @param _lendingPool      Address of the LendingPool contract
+     * @param _asset            The address of the underlying asset borrowed
+     * @param _rateMode         The rate mode that the user wants to swap to
+     */
+    function invokeSwapBorrowRateMode(
+        ISetToken _setToken,
+        IPool _lendingPool,
+        address _asset,
+        uint256 _rateMode
+    )
+        external
+    {
+        ( , , bytes memory callData) = getSwapBorrowRateModeCalldata(
+            _lendingPool,
+            _asset,
+            _rateMode
+        );
+        
+        _setToken.invoke(address(_lendingPool), 0, callData);
+    }
+
+    /**
+     * Invoke set Efficiency Mode
+     *
+     * Allows SetToken to borrow a specific `_amountNotional` of the reserve underlying `_asset`, provided that 
+     * the SetToken already deposited enough collateral, or it was given enough allowance by a credit delegator
+     * on the corresponding debt token (StableDebtToken or VariableDebtToken)
+     * @param _setToken             Address of the SetToken
+     * @param _lendingPool          Address of the LendingPool contract
+     * @param _categoryId           TODO: What is this ?
+     */
+    function invokeSetUserEMode(
+        ISetToken _setToken,
+        IPool _lendingPool,
+        uint8 _categoryId
+    )
+        external
+    {
+        ( , , bytes memory borrowCalldata) = getSetUserEmodeCalldata(
+            _lendingPool,
+            _categoryId
+        );
+        
+        _setToken.invoke(address(_lendingPool), 0, borrowCalldata);
+    }
+
+    function getSetUserEmodeCalldata(
+        IPool _lendingPool,
+        uint8 _categoryId
+    )
+        public
+        pure
+        returns (address, uint256, bytes memory)
+    {
+        bytes memory callData = abi.encodeWithSignature(
+            "setUserEMode(uint8)", 
+            _categoryId
+        );
+        
+        return (address(_lendingPool), 0, callData);
+    }
```

Comment:
you provide a comment on this new functionality that wasn't in the V2 Module?

what does it enable? how does it interact with other parts of the module? anything reviewers should key in on?

Reply:
So this allows setting the E-Mode ( a new feature of AaveV3) category on behalf of the SetToken.
Basically this allows us to increase the leverage ratio for tokens like icEth that have "like to like" collateral and borrow tokens by increasing LTV and decreasing liquidation penalty. (In this case weth and stEth are both in the same e-mode category).

After setting the E-Mode to a specific category (such as the "Eth" or "stablecoin" category), the user (in this case the set token instance) can only borrow assets from that category albeit with improved risk parameters.

Since the E-Mode changes the liquidation parameters it should be done with care. Especially if at some point we change to an E-Mode category that decreases LTV and thereby moves us closer to liquidation territory. (If I understand [the aave docs](https://docs.aave.com/developers/whats-new/efficiency-mode-emode#how-it-works) correctly it should revert if the e-mode change would result in undercollateralised positions).

Getting a good understanding of E-Mode and considering how / for which products we might want to use it would be a good avenue of thought.

A few resources on e-mode:
- [AaveV3 technical paper (Section 3.1)](https://github.com/aave/aave-v3-core/blob/master/techpaper/Aave_V3_Technical_Paper.pdf)

Reply:
Also slightly extended the docstring to elaborate on this topic somewhat.

Reply:
thanks for this @ckoopmann 👏

this new functionality LGTM

it'd be great to see some integration tests in E-Mode, and potentially compare some scenarios to non E-Mode tests. these tests may be better paired with an Aave V3 icETH deployment script.